### PR TITLE
cpu/esp8266: fix stdio problems

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR fixes a problem with the standard output that arose in conjunction with the `i2c_scan` problem described in issue #12125.

The problem occurs when stdio functions `puts`, `putchar` and `printf` are used together with `fputs` or `fprintf` and `stdout` as file parameter. While outputs with `puts` are printed out immediately, 
outputs with `fputs` are not printed out on linefeed but when the internal buffer of 128 bytes is full. As a result, mixing these functions scrambles the output completely.

The problem is fixed by removing the overridden stdio functions `puts`, `putchar` and `printf`. Instead, the corresponding newlib functions are used always. For this purpose, modules `newlib`, `newlib_syscalls_default` and `stdio_uart` are now enabled by default which also reduces the number of module dependency rules in `Makefile.dep`.

### Testing procedure

Flash and test `examples/default` with enabled module `i2c_scan`  with command
```
USEMODULE='i2c_scan' make BOARD=esp8266-esp-12x -C examples/default flash term
```
Without this PR the output should look like:
```
2019-08-31 14:17:35,993 - INFO #  i2c_scan 0
2019-08-31 14:17:35,996 - INFO # Scanning I2C device 0...
2019-08-31 14:17:36,002 - INFO # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2019-08-31 14:17:36,006 - INFO #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2019-08-31 14:17:36,006 - INFO # 
2019-08-31 14:17:36,007 - INFO # 
2019-08-31 14:17:36,007 - INFO # 
2019-08-31 14:17:36,019 - INFO # 0x00 R R R R R R R R R R R R R R E E0x10 E E E E E E E E E E E E E E E E0x20 E E E E E E E E E E E E E E E E0x30 E E E E E E E E
2019-08-31 14:17:36,020 - INFO # 
2019-08-31 14:17:36,020 - INFO # 
2019-08-31 14:17:36,021 - INFO # 
2019-08-31 14:17:36,030 - INFO #  E E E E E E E E0x40 E E E E E X E E E E E E E E E E0x50 E E E E E E E E E E E E E E E E0x60 E E E E E E E E E E E E E E E E0x70
```
With this PR the output should look like:
```
> i2c_scan 0
2019-08-31 14:21:33,249 - INFO #  i2c_scan 0
2019-08-31 14:21:33,252 - INFO # Scanning I2C device 0...
2019-08-31 14:21:33,257 - INFO # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2019-08-31 14:21:33,260 - INFO #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2019-08-31 14:21:33,263 - INFO # 0x00 R R R R R R R R R R R R R R E E
2019-08-31 14:21:33,266 - INFO # 0x10 E E E E E E E E E E E E E E E E
2019-08-31 14:21:33,271 - INFO # 0x20 E E E E E E E E E E E E E E E E
2019-08-31 14:21:33,274 - INFO # 0x30 E E E E E E E E E E E E E E E E
2019-08-31 14:21:33,277 - INFO # 0x40 E E E E E X E E E E E E E E E E
2019-08-31 14:21:33,280 - INFO # 0x50 E E E E E E E E E E E E E E E E
2019-08-31 14:21:33,283 - INFO # 0x60 E E E E E E E E E E E E E E E E
2019-08-31 14:21:33,287 - INFO # 0x70 E E E E E E E E R R R R R R R R
```

### Issues/PRs references

Related to issue #12125.

### CODE REVIEW

**Comment by benpicco** [comment]
Uh, but Murdock is unhappy

```
Test: comparing addresses of 'printf' and 'iprintf' symbols
[FAILED] '40101cf8' != '40101cf8' is False
Makefile:34: recipe for target 'test-newlib' failed
```

I guess `tests/libc_newlib` needs an update.

**Comment by gschorcht** [comment]
@benpicco Thanks for testing.

Unfortunatly, Murdock is not happy yet. There are two apps that aren't compilable anymore:
- `tests/libc_newlib`: Compilation fails since `printf` and `iprintf` have same address. I don't know at the moment, what the problem is if `iprintf` is simply mapped to `printf`.
- `tests/pkg_libb2`: Undefined references since these functions get an suffix when they are compiled, for example `blake2bSUFFIX`. I have no idea why.

**Comment by gschorcht** [comment]
> ```
> Test: comparing addresses of 'printf' and 'iprintf' symbols
> [FAILED] '40101cf8' != '40101cf8' is False
> Makefile:34: recipe for target 'test-newlib' failed
> ```

It is not clear for me why the test complains that addresses are equal. I'm a bit unsure if it would be the right way to blacklist ESP8266 for this test.

Maybe @cladmi can say more about this test and whether it is really a problem if it fails.

**Comment by benpicco** [comment]
`tests/pkg_libb2` also fails on master with the same error, so this is not related to this PR.

**Comment by cladmi** [comment]
> > ```
> > Test: comparing addresses of 'printf' and 'iprintf' symbols
> > [FAILED] '40101cf8' != '40101cf8' is False
> > Makefile:34: recipe for target 'test-newlib' failed
> > ```
> 
> It is not clear for me why the test complains that addresses are equal. I'm a bit unsure if it would be the right way to blacklist ESP8266 for this test.
> 
> Maybe @cladmi can say more about this test and whether it is really a problem if it fails.

The goal of this test was to detect if `newlib-nano` is indeed used when requested.

I had no real idea how to do the detection so used the fact described in the
`newlib-nano` that `iprintf` is mapped to `printf`

https://github.com/RIOT-OS/RIOT/blob/d991883c5b84ae1d9d169131156edade1cd67f6f/tests/libc_newlib/Makefile#L23-L31

However, it appears that it is not a reliable condition :'(

Could it be that `esp` is using `newlib-nano` and not `newlib` ?

I will try debugging a bit.

**Comment by cladmi** [comment]
TL;DR: Test is right, you just need to add `USEMODULE += newlib_nano` after the `USEMODULE += newlib` to declare it in the build system, and verify everything else keeps working :D 

#### Explanation

According to this test the `esp8266` newlib is using the `newlib-nano` specs.
So the test is correct and showing an error in the configuration.

For some toolchain it must be explicitly enabled and handled in the build system but it looks like it is by default for `esp8266`. However, we rely on having the `newlib_nano` to be in `USEMODULE`.


To verify this, I faked enabling newlib_nano in my test, and it detects `newlib_nano`:

<details><summary>Configure the test to expect `newlib_nano`</summary>

```diff
diff --git a/tests/libc_newlib/Makefile b/tests/libc_newlib/Makefile
index f21f0761d..f3a4e781c 100644
--- a/tests/libc_newlib/Makefile
+++ b/tests/libc_newlib/Makefile
@@ -2,6 +2,8 @@ include ../Makefile.tests_common
 
 USEMODULE += embunit
 
+CFLAGS += -DMODULE_NEWLIB_NANO=1
+
 include $(RIOTBASE)/Makefile.include
 
 # Compile time tests
@@ -20,6 +22,9 @@ ifneq (,$(filter newlib,$(USEMODULE)))
   endif
 endif
 
+# Force trying 'NEWLIB_NANO'
+CMP_OP = =
+
 # Newlib-nano removed the integer only 'iprintf' functions which are now mapped
 # to printf.
 #
```
</details>

<details><summary>BUILD_IN_DOCKER=1 DOCKER="sudo docker" BOARD=esp8266-esp-12x make flash term</summary>

```
...
Test: comparing addresses of 'printf' and 'iprintf' symbols
[SUCCESS] '40101cf8' = '40101cf8' is True
...
2019-09-02 12:15:11,712 - INFO # sysmem: 5116 (used 612, free 4504)
2019-09-02 12:15:11,712 - INFO # 
2019-09-02 12:15:11,713 - INFO # Board configuration:
2019-09-02 12:15:11,715 - INFO # 	PWM_DEV(0): channelmain(): This is RIOT! (Version: 2019.10-devel-559-g39984c-pr/riot/12133/fix_stdio_problems)
2019-09-02 12:15:11,715 - INFO # Newlib/nano test
2019-09-02 12:15:11,715 - INFO # .
2019-09-02 12:15:11,715 - INFO # OK (1 tests)
2019-09-02 12:15:57,686 - INFO # Exiting Pyterm
```
</details>

So this correctly compiles and detects the `_NANO_FORMATTED_IO` macro.

https://github.com/RIOT-OS/RIOT/blob/3901740ef219389d63ff859213f3c3b22c0ec45f/tests/libc_newlib/main.c#L39-L43


<details><summary>Adding `newlib_nano` to `cpu/esp8266` correctly makes this test pass too without modifications</summary>

``` diff
diff --git a/cpu/esp8266/Makefile.include b/cpu/esp8266/Makefile.include
index 6f2a999c6..3104bd587 100644
--- a/cpu/esp8266/Makefile.include
+++ b/cpu/esp8266/Makefile.include
@@ -62,6 +62,7 @@ PSEUDOMODULES += esp_spiffs
 USEMODULE += esp
 USEMODULE += mtd
 USEMODULE += newlib
+USEMODULE += newlib_nano
 USEMODULE += newlib_syscalls_default
 USEMODULE += periph
 USEMODULE += ps
```
</details>

I love when tests can do reviews automatically :)

**Comment by gschorcht** [comment]
@cladmi Thanks for clarifying that. Indeed, the newlib-xtensa was compiled with `newlib-nano-formatted-io` enabled. Using module `newlib_nano` solves the problem. Thanks again.

**Comment by gschorcht** [comment]
@benpicco Murdock is happy now.

**Comment by benpicco** [comment]
Any my esp8266 is happy now too!

**Comment by gschorcht** [comment]
@benpicco Thanks again for reviewing and testing.
